File: Data Structures & Algorithms/three-integer-sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        set<vector<int>> uniqueTriplets; // to store unique sorted triplets
        int n = nums.size();

        for (int i = 0; i < n; ++i) {
            int target = -nums[i];
            unordered_set<int> seen;
            for (int j = i + 1; j < n; ++j) {
                int complement = target - nums[j];
                if (seen.count(complement)) {
                    vector<int> triplet = {nums[i], nums[j], complement};
                    sort(triplet.begin(), triplet.end()); // sort before inserting to handle permutations
                    uniqueTriplets.insert(triplet);
                }
                seen.insert(nums[j]);
            }
        }

        // Convert set to vector
        return vector<vector<int>>(uniqueTriplets.begin(), uniqueTriplets.end());
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        prevNum = math.inf
        finalAns = list(list())
        nums.sort()

        for i in range(0,len(nums)) :
            left = i + 1 
            right = len(nums) - 1

            if prevNum == nums[i] :
                continue 

            prevNum = nums[i]

            while left < right :
                sumation = nums[i] + nums[left] + nums[right] 

                if sumation == 0 and (nums[i], nums[left] , nums[right]) not in finalAns :
                    finalAns.append((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1
                    

                elif sumation > 0 :
                    right -= 1

                else :
                    left += 1


        return finalAns